actors/miner/src/lib.rs

```diff
@@ -1010,190 +941,234 @@ impl Actor {
+        }
 
-            // Update pledge.
-            let current_balance = rt.current_balance();
-            if pledge_delta.is_positive() {
-                let unlocked_balance =
-                    state.get_unlocked_balance(&current_balance).map_err(|e| {
-                        actor_error!(illegal_state, "failed to calculate unlocked balance: {}", e)
-                    })?;
-                if unlocked_balance < pledge_delta {
-                    return Err(actor_error!(
-                    insufficient_funds,
-                    "insufficient funds for aggregate initial pledge requirement {}, available: {}",
-                    pledge_delta,
-                    unlocked_balance
-                ));
-                }
-            }
+        // Successful data activation is required for sector activation.
+        let successful_manifests = data_batch.successes(&proven_manifests);
 
-            state
-                .add_initial_pledge(&pledge_delta)
-                .map_err(|e| actor_error!(illegal_state, "failed to add initial pledge: {}", e))?;
+        let mut state_updates_by_dline = BTreeMap::<u64, Vec<ReplicaUpdateStateInputs>>::new();
+        for ((update, sector_info), data_activation) in
+            successful_manifests.iter().zip(data_activations)
+        {
+            let activated_data = ReplicaUpdateActivatedData {
+                seal_cid: update.new_sealed_cid,
+                unverified_space: data_activation.unverified_space.clone(),
+                verified_space: data_activation.verified_space.clone(),
+            };
+            state_updates_by_dline.entry(update.deadline).or_default().push(
+                ReplicaUpdateStateInputs {
+                    deadline: update.deadline,
+                    partition: update.partition,
+                    sector_info,
+                    activated_data,
+                },
+            );
+        }
 
-            state.check_balance_invariants(&current_balance).map_err(balance_invariants_broken)?;
-            Ok(())
-        })?;
+        let (power_delta, pledge_delta) = update_replica_states(
+            rt,
+            &state_updates_by_dline,
+            successful_manifests.len(),
+            &mut sectors,
+            info.sector_size,
+        )?;
 
         notify_pledge_changed(rt, &pledge_delta)?;
         request_update_power(rt, power_delta)?;
 
-        let updated_bitfield = BitField::try_from_bits(updated_sectors)
-            .context_code(ExitCode::USR_ILLEGAL_ARGUMENT, "invalid sector number")?;
-        Ok(updated_bitfield)
+        // Notify data consumers.
+        let mut notifications: Vec<ActivationNotifications> = vec![];
+        for (update, sector_info) in successful_manifests {
+            notifications.push(ActivationNotifications {
+                sector_number: update.sector,
+                sector_expiration: sector_info.expiration,
+                pieces: &update.pieces,
+            });
+        }
```

Comment:
nit:

```suggestion
        let notifications: Vec<ActivationNotifications> = successful_manifests
            .into_iter()
            .map(|(update, sector_info)| ActivationNotifications {
                sector_number: update.sector,
                sector_expiration: sector_info.expiration,
                pieces: &update.pieces,
            })
            .collect();
```

(for simple cases, not really critical I guess).